toolchain/check/facet_type.cpp

```diff
@@ -263,4 +251,91 @@ auto AllocateFacetTypeImplWitness(Context& context,
+    return false;
+  }
+  const auto& symbolic =
+      context.constant_values().GetSymbolicConstant(const_id);
+  // Fast early reject before doing more expensive operations.
+  if (symbolic.dependence != SemIR::ConstantDependence::PeriodSelf) {
+    return false;
+  }
+  return IsPeriodSelf(context, symbolic.inst_id);
+}
+
+auto IsPeriodSelf(Context& context, SemIR::InstId inst_id) -> bool {
+  // Unwrap the `FacetAccessType` instruction, which we get when the `.Self` is
+  // converted to `type`.
+  if (auto facet_access_type =
+          context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
+    inst_id = facet_access_type->facet_value_inst_id;
+  }
+  if (auto bind_symbolic_name =
+          context.insts().TryGetAs<SemIR::BindSymbolicName>(inst_id)) {
+    const auto& bind_name =
+        context.entity_names().Get(bind_symbolic_name->entity_name_id);
+    return bind_name.name_id == SemIR::NameId::PeriodSelf;
+  }
+  return false;
+}
+
+auto ResolveRewriteConstraintsAndCanonicalize(Context& context,
+                                              SemIR::LocId loc_id,
+                                              SemIR::FacetTypeInfo& facet_type)
+    -> void {
+  // This operation sorts and dedupes the rewrite constraints. They are sorted
+  // primarily by the `lhs_id`, then by the `rhs_id`.
+  facet_type.Canonicalize();
+
+  if (facet_type.rewrite_constraints.empty()) {
+    return;
+  }
+
+  for (size_t i :
+       llvm::seq<size_t>(facet_type.rewrite_constraints.size() - 1)) {
+    auto& constraint = facet_type.rewrite_constraints[i];
+    if (constraint.lhs_id == SemIR::ErrorInst::InstId ||
+        constraint.rhs_id == SemIR::ErrorInst::InstId) {
+      continue;
+    }
+
+    if (!context.insts().Is<SemIR::ImplWitnessAccess>(constraint.lhs_id)) {
+      continue;
+    }
+    auto lhs_access =
+        context.insts().GetAs<SemIR::ImplWitnessAccess>(constraint.lhs_id);
+    if (!context.insts().Is<SemIR::LookupImplWitness>(lhs_access.witness_id)) {
+      continue;
+    }
+    auto lhs_lookup =
+        context.insts().GetAs<SemIR::LookupImplWitness>(lhs_access.witness_id);
+    if (!IsPeriodSelf(context, lhs_lookup.query_self_inst_id)) {
+      continue;
+    }
+
+    auto& next = facet_type.rewrite_constraints[i + 1];
+    if (next.lhs_id != SemIR::ErrorInst::InstId &&
+        next.rhs_id != SemIR::ErrorInst::InstId) {
+      if (constraint.lhs_id == next.lhs_id) {
+        CARBON_DIAGNOSTIC(
+            AssociatedConstantWithDifferentValues, Error,
+            "associated constant {0} given two different values {1} and {2}",
+            InstIdAsConstant, InstIdAsConstant, InstIdAsConstant);
+        context.emitter().Emit(loc_id, AssociatedConstantWithDifferentValues,
+                               constraint.lhs_id, constraint.rhs_id,
+                               next.rhs_id);
+        constraint.rhs_id = SemIR::ErrorInst::InstId;
+        next.rhs_id = SemIR::ErrorInst::InstId;
+        continue;
+      }
+    }
```

Comment:
Combine `if` here?

```suggestion
    auto& next = facet_type.rewrite_constraints[i + 1];
    if (next.lhs_id != SemIR::ErrorInst::InstId &&
        next.rhs_id != SemIR::ErrorInst::InstId &&
        constraint.lhs_id == next.lhs_id) {
      CARBON_DIAGNOSTIC(
          AssociatedConstantWithDifferentValues, Error,
          "associated constant {0} given two different values {1} and {2}",
          InstIdAsConstant, InstIdAsConstant, InstIdAsConstant);
      context.emitter().Emit(loc_id, AssociatedConstantWithDifferentValues,
                             constraint.lhs_id, constraint.rhs_id,
                             next.rhs_id);
      constraint.rhs_id = SemIR::ErrorInst::InstId;
      next.rhs_id = SemIR::ErrorInst::InstId;
      continue;
    }
```

Reply:
Thinking about this further, maybe you could change the `if` a little since we should already know validity of `constraint.lhs_id`:

```suggestion
    auto& next = facet_type.rewrite_constraints[i + 1];
    if (next.lhs_id == constraint.lhs_id &&
        next.rhs_id != SemIR::ErrorInst::InstId) {
      CARBON_DIAGNOSTIC(
          AssociatedConstantWithDifferentValues, Error,
          "associated constant {0} given two different values {1} and {2}",
          InstIdAsConstant, InstIdAsConstant, InstIdAsConstant);
      context.emitter().Emit(loc_id, AssociatedConstantWithDifferentValues,
                             constraint.lhs_id, constraint.rhs_id,
                             next.rhs_id);
      constraint.rhs_id = SemIR::ErrorInst::InstId;
      next.rhs_id = SemIR::ErrorInst::InstId;
      continue;
    }
```

Reply:
Done